dallinger/config.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import unicode_literals
+
+from collections import deque
+from ConfigParser import SafeConfigParser
+import distutils.util
+import os
+import threading
+
+marker = object()
+
+try:
+    unicode = unicode
+except NameError:  # Python 3
+    unicode = str
+
+
+class Configuration(object):
+
+    SUPPORTED_TYPES = {
+        bytes,
+        unicode,
+        int,
+        float,
+        bool,
+    }
+
+    def __init__(self):
+        self.data = deque()
+        self.types = {}
+        self.synonyms = {}
+        self.ready = False
+
+    def extend(self, mapping, cast_types=False):
+        normalised_mapping = {}
+        for key, value in mapping.items():
+            key = self.synonyms.get(key, key)
+            if key not in self.types:
+                # This key hasn't been registered, we ignore it
+                continue
+            expected_type = self.types.get(key)
+            if cast_types:
+                try:
+                    if expected_type == bool:
+                        value = distutils.util.strtobool(value)
+                    value = expected_type(value)
+                except ValueError:
+                    pass
+            if not isinstance(value, expected_type):
+                raise ValueError(
+                    'Got {value} for {key}, expected {expected_type}'
+                    .format(
+                        value=repr(value),
+                        key=key,
+                        expected_type=expected_type,
+                    )
+                )
+            normalised_mapping[key] = value
+        self.data.extendleft([normalised_mapping])
+
+    def get(self, key, default=marker):
+        if not self.ready:
+            raise RuntimeError('Config loading not finished')
+        for layer in self.data:
+            try:
+                return layer[key]
+            except KeyError:
+                continue
+        if default is marker:
+            raise KeyError(key)
+        return default
+
+    def register(self, key, type_, synonyms=set()):
+        if type_ not in self.SUPPORTED_TYPES:
+            raise ValueError(
+                '{type} is not a supported type'.format(
+                    type=type_
+                )
+            )
+        self.types[key] = type_
+        for synonym in synonyms:
+            self.synonyms[synonym] = key
+
+    def load_from_config_file(self, filename):
+        parser = SafeConfigParser()
+        parser.read(filename)
+        data = {}
+        for section in parser.sections():
+            data.update(dict(parser.items(section)))
```

Comment:
I did ask for thoughts about this on Slack as I thought it might be controversial. The conversation was on the 15th of December if you want to look back at it.

The inspiration was taken from the Django configuration, with meaningful single keys, rather than a compound key of section/value. There are some advantages in that you can represent such keys as command line arguments or environment variables easily, and the code is more readable (in my opinion), but if it's useful we can keep the sections!

As to your example about debug mode, I guess you're suggesting something like:

```
[Live Server Parameters]
dyno_type = performance-m
num_dynos_web = 4
num_dynos_worker = 2
host = 0.0.0.0
notification_url = None
clock_on = true
logfile = -

[Debug Server Parameters]
dyno_type = free
num_dynos_web = 1
num_dynos_worker = 1
host = 0.0.0.0
notification_url = None
clock_on = false
logfile = -

[Server Parameters]
dyno_type = ${Debug Server Parameters:dyno_type}
num_dynos_web = ${Debug Server Parameters:num_dynos_web}
num_dynos_worker = ${Debug Server Parameters:num_dynos_worker}
host = ${Debug Server Parameters:host}
notification_url = ${Debug Server Parameters:notification_url}
clock_on = ${Debug Server Parameters:clock_on}
logfile = ${Debug Server Parameters:logfile}
```

but that feels rather ugly to me. I would suggest that this would be better achieved by supporting, for example, a debug-overrides.ini file that specifies overrides for debug mode.